### PR TITLE
[ome_configuration_compliance_baseline] - Added support for scheduling or staging the job

### LINKED ISSUE #616: [FEATURE]: ome_configuration_compliance_baseline stage on next reboot or schedule job
the ome_configuration_compliance_baseline module could be extended.

when remediating a configuration baseline, the job is executed immediately.
this creates problems when a reboot is included. 

In OME it is possible to 

- schedule the job at a specific time or
- to stage and execute the job at the next reboot

maybe:

```
- name: Remediate specific non-compliant devices to a configuration compliance baseline using device service tags
  dellemc.openmanage.ome_configuration_compliance_baseline:
    hostname: "192.168.0.1"
    username: "username"
    password: "password"
    ca_path: "/path/to/ca_cert.pem"
    command: "remediate"
    names: "baseline1"
    device_service_tags:
      - "SVCTAG1"
      - "SVCTAG2"
    schedule: stage
    schedule_intervall: next_reboot

- name: Remediate specific non-compliant devices to a configuration compliance baseline using device service tags
  dellemc.openmanage.ome_configuration_compliance_baseline:
    hostname: "192.168.0.1"
    username: "username"
    password: "password"
    ca_path: "/path/to/ca_cert.pem"
    command: "remediate"
    names: "baseline1"
    device_service_tags:
      - "SVCTAG1"
      - "SVCTAG2"
    schedule: stage
    schedule_intervall: once
    scheduleStart: "2024-03-01T10:00:00.000Z"
```

### PR DESCRIPTION
**Pull request by ShivamSh3**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 1]
```diff
@@ -745,23 +794,62 @@ def create_remediate_payload(noncomplaint_devices, baseline_info, rest_obj):
             "RunLater": False
         }
     }
+    if module.params.get("run_later"):
+        if validate_cron(module.params.get("cron")) is False:
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/ome_configuration_compliance_baseline.py]: use `not` to check the negative boolean condition.

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 2]
```diff
@@ -745,23 +794,62 @@ def create_remediate_payload(noncomplaint_devices, baseline_info, rest_obj):
...
+            module.exit_json(msg=INVALID_CRON_TIME_MSG, failed=True)
+        payload = {
+            "Id": baseline_info["Id"],
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/ome_configuration_compliance_baseline.py]: use `.get()` here.

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 3]
```diff
@@ -745,23 +794,62 @@ def create_remediate_payload(noncomplaint_devices, baseline_info, rest_obj):
...
+            "Schedule": {
+                "RunNow": False,
+                "RunLater": True,
+                "Cron": module.params.get("cron")
+            }
+        }
+    elif module.params.get("staged_at_reboot"):
+        payload = {
+            "Id": baseline_info["Id"],
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/ome_configuration_compliance_baseline.py]: same as above comment.

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 4]
```diff
@@ -745,23 +794,62 @@ def create_remediate_payload(noncomplaint_devices, baseline_info, rest_obj):
...
     if LooseVersion(ome_version) >= "3.5":
         payload["DeviceIds"] = noncomplaint_devices
     else:
         payload["TargetIds"] = noncomplaint_devices
     return payload
 
 
+def validate_cron(cron_string):
+    cron_pattern = CRON_REGEX
+    if re.match(cron_pattern, cron_string) is not None:
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/ome_configuration_compliance_baseline.py]: This snippet can be enhanced.

`return bool(re.match(cron_pattern, cron_string))`

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 1 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: done

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 2 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: done

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 3 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: Done

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 4 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: Done

File: plugins/module_utils/ome.py [hunk 5]
```diff
@@ -406,6 +408,24 @@ def _get_omam_ca_env(self):
...
+            tmp_dict = dict((x["StartTime"], x["Id"]) for x in ex_hist)
+            sorted_dates = sorted(tmp_dict.keys())
+            ex_url = JOB_EXEC_HISTORY.format(job_id=job_id) + "({0})/ExecutionHistoryDetails".format(tmp_dict[sorted_dates[-1]])
+            all_exec = self.get_all_items_with_pagination(ex_url)
+            for jb_ip in all_exec.get('value'):
+                jb_ip = strip_substr_dict(jb_ip)
+                jb_ip.get('JobStatus', {}).pop('@odata.type', None)
+                job_detail_status.append(jb_ip)
+        except Exception:
+            pass
```
Comment (jagadeeshnv) [inline plugins/module_utils/ome.py:426]: resutrn empty list or some accurate message

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 6]
```diff
@@ -736,32 +788,66 @@ def validate_remediate_idempotency(module, rest_obj):
...
     resp = rest_obj.invoke_request('POST', REMEDIATE_BASELINE, data=remediate_payload)
     job_id = resp.json_data
+    if module.params.get("run_later"):
+        schedule_job(module, rest_obj, job_id)
     if module.params.get("job_wait"):
         job_failed, message = rest_obj.job_tracking(job_id, job_wait_sec=module.params["job_wait_timeout"])
         if job_failed is True:
-            module.fail_json(msg=message, job_id=job_id, changed=False)
+            detailed_output = rest_obj.get_job_execution_details(job_id)
+            module.exit_json(msg=message, job_id=job_id, job_details=detailed_output, failed=True)
```
Comment (jagadeeshnv) [inline plugins/modules/ome_configuration_compliance_baseline.py:872]: Add this key job_details in the above doc "RETURN" section

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 7]
```diff
@@ -93,6 +93,21 @@
...
     type: str
+  run_later:
+    description:
+      - Indicates whether template is to be assigned immediately or in the future.
+      - If I(run_later) is C(true), then I(staged_at_reboot) is ignored.
+      - If I(run_later) is C(true), then I(job_wait) is not applicable.
+    type: bool
+  cron:
+    description:
+      - Provide a cron expression based on Quartz cron format.
```
Comment (jagadeeshnv) [inline plugins/modules/ome_configuration_compliance_baseline.py:106]: Describe the format - and also mention it is a specific datetime

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 8]
```diff
@@ -93,6 +93,21 @@
       - Description of the compliance baseline.
       - This option is applicable when I(command) is C(create), or C(modify).
     type: str
+  run_later:
+    description:
+      - Indicates whether template is to be assigned immediately or in the future.
```
Comment (rajshekarp87) [inline plugins/modules/ome_configuration_compliance_baseline.py]:
```suggestion
      - Remediate immediately or at next reboot. This is applicable when I(command) is C(remediate).
```

**Comment by rajshekarp87** [comment]
Add job_details sample output in RETURN

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 7 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py:106]: done

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 6 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py:872]: done

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 8 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: done

**Review by rajshekarp87 — DISMISSED** [review]

**Review by rajshekarp87 — DISMISSED** [review]

File: plugins/modules/ome_configuration_compliance_baseline.py [hunk 9]
```diff
@@ -266,6 +312,27 @@
...
+            "ElapsedTime": "00:22:17",
+            "EndTime": "2024-06-19 13:42:41.285",
+            "ExecutionHistoryId": 797320,
+            "Id": 14123,
+            "IdBaseEntity": 19559,
+            "JobStatus": {
+                "Id": 2070,
+                "Name": "Failed"
+            },
+            "Key": "7D0K7Y3",
```
Comment (jagadeeshnv) [inline plugins/modules/ome_configuration_compliance_baseline.py]: Repalce with dummy value

**Review by jagadeeshnv — DISMISSED** [review]

File: plugins/modules/ome_configuration_compliance_baseline.py (hunk 9 above)
Comment (ShivamSh3) [reply plugins/modules/ome_configuration_compliance_baseline.py]: done

**Review by jagadeeshnv — APPROVED** [review]

**Review by ABHISHEK-SINHA10 — APPROVED** [review]

**Review by sachin-apa — APPROVED** [review]

**Pull request merged**